osu.Game/Database/RealmAccess.cs

```diff
@@ -910,6 +930,9 @@ void restoreOperation()
 
         public void Dispose()
         {
+            if (!pendingAsyncWrites.Wait(10000))
+                Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
```

Comment:
This lock ordering is giving me the creeps. `WriteAsync()` takes `realmLock` first and then adds count to `pendingAsyncWrites`, while this waits on the latter first and only ever then takes the former. Not to mention that `pendingAsyncWrites` is being signaled from TPL. The timeout ensures that this won't deadlock, but I don't know that this will never timeout, nor that it will work properly.

At least in my head, if this `.Wait()` is done after taking `realmLock`, then there's the guarantee that no new writers can add a count to the `CountdownEvent` (because they'd have to take `realmLock` first).

Reply:
Not sure I see it. If I apply

```diff
diff --git a/osu.Game/Database/RealmAccess.cs b/osu.Game/Database/RealmAccess.cs
index 7a9561fc5e..e3edec254b 100644
--- a/osu.Game/Database/RealmAccess.cs
+++ b/osu.Game/Database/RealmAccess.cs
@@ -930,11 +930,11 @@ void restoreOperation()
 
         public void Dispose()
         {
-            if (!pendingAsyncWrites.Wait(10000))
-                Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
-
             lock (realmLock)
             {
+                if (!pendingAsyncWrites.Wait(10000))
+                    Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
+
                 updateRealm?.Dispose();
             }
```

all test cases from files touched in this diff still pass.

Reply:
Probably just good luck. Have a look at `WriteAsync` -> `getRealmInstance` -> `realmLock`

I’ll try and add better test coverage of this tomorrow…

Reply:
I know this is going to likely increase the complexity of this PR, but I've removed `realmLock` completely. I can't understand why it would be required. I've also run tests through quite a few iterations with no failures.

Let me know your thoughts. Probably worth taking a good look through the class and making sure I'm not missing something. If we need more locking than the `realmRetrievalLock`, it needs to be named and documented far better.

Reply:
Right now, as I see it, the primary things that `realmLock` is protecting are probably the `updateRealm` singleton and the `customSubscriptionResetMap`, which is a non-concurrent collection. Most private places are guaranteed to be taking that lock on the update thread (`ensureUpdateRealm()`, `RegisterCustomSubscription+unsubscribe()`, which is sent/posted, `registerSubscription()` which is either called from `ensureUpdateRealm()`, or sent/posted).

The ones that worry me more are the public ones: `RegisterForNotifications<T>()`, `BlockAllOperations()`, `Dispose()`. They *look* like there should be no thread contention on them (always should be getting called from update), but I'm not 100% sure if I'm honest. Maybe fine to stick `ThreadSafety.IsUpdateThread` checks everywhere and call it a day?

Reply:
Yeah, I think enforcing update thread is much better than locking.. until we need otherwise.